core/src/main/java/jenkins/util/SetContextClassLoader.java

```diff
@@ -0,0 +1,74 @@
+import java.io.ObjectInputStream;
+
+/**
+ * Java defines a {@link Thread#contextClassLoader}. Jenkins does not use this much; it will
+ * normally be set by the servlet container to the Jenkins core class loader.
+ *
+ * <p>Some Java libraries have a fundamental design flaw, originating in premodular systems with a
+ * "flat classpath", whereby they expect {@link Thread#contextClassLoader} to have access to the
+ * same classes as the class loader of the calling class. This fails in Jenkins, because {@link
+ * Thread#contextClassLoader} can only see Jenkins core, not plugins.
+ *
+ * <p>It is a design flaw in the library if it fails to allow clients to directly specify a {@link
+ * ClassLoader} to use for lookups (or preregister {@link Class} instances for particular names).
+ * Consider patching the library or looking harder for appropriate APIs that already exist. As an
+ * example, {@link ObjectInputStream} (used for deserializing Java objects) by default uses a
+ * complicated algorithm to guess at a {@link ClassLoader}, but you can override {@link
+ * ObjectInputStream#resolveClass} to remove the need for guessing (as {@link ObjectInputStreamEx}
+ * in fact does).
+ *
+ * <p>Alternatively, work around the problem by applying {@link SetContextClassLoader} liberally in
+ * a {@code try}-with-resources block wherever we might be calling into such a library:
+ *
+ * <pre>
+ * class Caller {
+ *     void foo() {
+ *         try (SetContextClassLoader sccl = new SetContextClassLoader(Caller.class)) {
+ *             [...] // Callee uses Thread.currentThread().getContextClassLoader()
+ *         }
+ *     }
+ * }
+ * </pre>
+ *
+ * <p>When called from a plugin, {@link #SetContextClassLoader(Class)} should typically be used, its
+ * argument being the calling class, whose class loader has access to all the plugin's direct and
+ * transitive dependencies. When the particular class loader needed is unclear, {@link
+ * #SetContextClassLoader(ClassLoader)} can be used as a fallback with {@link
+ * PluginManager.UberClassLoader} as the argument, though this is not as safe since lookups could be
+ * ambiguous in case two unrelated plugins both bundled the same library. In functional tests,
+ * {@code RealJenkinsRule.Endpoint} can be used to reference a class loader that has access to the
+ * plugins defined in the test scenario.
+ *
+ * <p>See <a
+ * href="https://www.jenkins.io/doc/developer/plugin-development/dependencies-and-class-loading/#context-class-loaders">the
+ * developer documentation</a> for more information.
+ *
+ * @since TODO
+ */
+public final class SetContextClassLoader implements AutoCloseable {
+
+    private final Thread t;
+    private final ClassLoader orig;
+
+    public SetContextClassLoader(Class<?> clazz) {
+        t = Thread.currentThread();
+        orig = t.getContextClassLoader();
+        // It is too bad that Reflection.getCallerClass() is a private API.
+        t.setContextClassLoader(clazz.getClassLoader());
+    }
```

Comment:
Once we're on Java 11, we could perhaps remove the `clazz` argument and get the calling class via the [`StackWalker`](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/lang/StackWalker.html) API.

Reply:
FWIW I would much prefer the explicit `Class` argument as currently written here. It is hardly bothersome to pass, and sometimes you may actually want something other than the calling class: a class from a library, a plugin you depend upon, etc. Magically inferring an argument according to the caller is a poor practice in terms of legibility and predictability; for example, it means that supposedly safe refactorings like pull up method may quietly lead to different behavior.

Reply:
Well you can have your opinion and I can have mine. I actually think a no-argument version is preferable since you can cut and paste code from one class to another without having to worry about updating the calling class argument to reflect the new home of the code.